3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/IabEnrichment.scala

```diff
@@ -0,0 +1,265 @@
+  private def getDatabaseUri(uri: String, database: String): ValidatedMessage[URI] =
+    ConversionUtils
+      .stringToUri(uri + "/" + database)
+      .flatMap {
+        case Some(u) => u.success
+        case None    => "URI to IAB file must be provided".fail
+      }
+      .toProcessingMessage
+}
+
+/**
+ * Contains enrichments based on IAB Spiders&Robots lookup.
+ *
+ * @param ipFileTuple        (Full URI to the IAB excluded IP list, database name)
+ * @param excludeUaFileTuple (Full URI to the IAB excluded user agent list, database name)
+ * @param includeUaFileTuple (Full URI to the IAB included user agent list, database name)
+ * @param httpUsername       Optional username for basic HTTP authentication
+ * @param httpPassword       Optional password for basic HTTP authentication
+ * @param localMode          Whether to use the local database file. Enabled for tests.
+ */
+case class IabEnrichment(
+  ipFileTuple: FileTuple,
+  excludeUaFileTuple: FileTuple,
+  includeUaFileTuple: FileTuple,
+  httpUsername: Option[String],
+  httpPassword: Option[String],
+  localMode: Boolean
+) extends Enrichment {
+
+  private type DbEntry = Option[(Option[URI], String)]
+
+  private val schemaUri        = "iglu:com.iab.snowplow/iab_spiders_and_robots/jsonschema/1-0-0"
+  private implicit val formats = DefaultFormats
+
+  // Construct a Tuple3 of all IAB files
+  private val dbs: (DbEntry, DbEntry, DbEntry) = {
+
+    def db(dbPath: FileTuple): DbEntry = dbPath.map {
+      case (name, uri, file) =>
+        if (localMode) {
+          (None, getClass.getResource(file).toURI.getPath)
+        } else {
+          (Some(uri), name)
+        }
+    }
+
+    (db(ipFileTuple), db(excludeUaFileTuple), db(includeUaFileTuple))
+  }
+
+  // Create an IAB client based on the IAB files list
+  private val iabClient = {
+    def file(db: DbEntry): File = new File(db.get._2)
+
+    new IabClient(file(dbs._1), file(dbs._2), file(dbs._3))
+  }
+
+  /**
+   * Get the IAB response containing information about whether an event is a
+   * spider or robot using the IAB client library.
+   *
+   * @param userAgent  User agent used to perform the check
+   * @param ipAddress  IP address used to perform the check
+   * @param accurateAt Date of the event, used to determine whether entries in the
+   *                   IAB list are relevant or outdated
+   * @return an IabResponse object
+   */
+  private[enrichments] def performCheck(userAgent: String,
+                                        ipAddress: String,
+                                        accurateAt: DateTime): Validation[Throwable, IabEnrichmentResponse] =
+    try {
+      val result = iabClient.checkAt(userAgent, InetAddress.getByName(ipAddress), accurateAt.toDate)
+      IabEnrichmentResponse(result.isSpiderOrRobot,
+                            result.getCategory.toString,
+                            result.getReason.toString,
+                            result.getPrimaryImpact.toString).success
```

Comment:
can we keep the types further? i.e. no `toString` here

Reply:
The reason this is converted into a scala case class is because those types of Java enums seem to have some issues when they're serialized via `Extraction.decompose()`. I could convert the `IabResponse` into an `IabEnrichmentResponse` in `getIab` instead but I don't think that would make much of a difference.